### PR TITLE
Feature/cart client

### PR DESCRIPTION
**Pull request by princepatoliya**
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

- **What is the current behavior?** (You can also link to an open issue here)

- **What is the new behavior (if this is a feature change)?**

- **Other information**:

### CODE REVIEW

File: src/client/CartClient.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    const url = `${APIConstants.API_BASE_URL}/carts`;
+
+    return this._get<CartData>(url, params, {
+      [APIConstants.ZECKO_ACCESS_TOKEN_HEADER_KEY]: this.accessToken,
+    });
+  }
+
+  /**
+   * @param {JSON} cartActionRequest CartActionRequest
```
Comment (akartik80) [inline src/client/CartClient.ts:39]: instead of `JSON`, define an interface and use that

File: src/client/CartClient.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+   * @property `customerId`
+   * ClientCustomer Id - string property of cartActionRequest
+   * @property `variantId`
+   * Product Variant Id - string property of cartActionRequest
```
Comment (akartik80) [inline src/client/CartClient.ts:43]: `string property of cartActionRequest` should not be needed if you use interface

File: src/client/CartClient.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+   * @property `quantity`
+   * Product Quantity - number property of cartActionRequest
+   *
+   * @returns {Promise<CartData>}
+   * Paginated cart list of maximum 250 cart items. To request further cart items, use `lineItemsAfter` parameter.
```
Comment (akartik80) [inline src/client/CartClient.ts:48]:
```suggestion
   * Cart object containing maximum 20 line items in cart. To request further line items, use `lineItemsAfter` parameter.
```

Comment (akartik80) [reply src/client/CartClient.ts:48]: Please make sure you verify limits in code before putting in docs

File: src/client/CartClient.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    const url = `${APIConstants.API_BASE_URL}/carts/add`;
+
+    return this._patch<CartData>(url, null, cartActionRequest, {
+      [APIConstants.ZECKO_ACCESS_TOKEN_HEADER_KEY]: this.accessToken,
+      'Content-Type': 'application/json',
+    });
+  }
+
+  /**
+   * @typedef {JSON} cartActionRequest CartActionRequest
```
Comment (akartik80) [inline src/client/CartClient.ts:60]: same comments here

File: src/client/CartClient.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    const url = `${APIConstants.API_BASE_URL}/carts/delete`;
+
+    return this._patch<CartData>(url, null, cartActionRequest, {
+      [APIConstants.ZECKO_ACCESS_TOKEN_HEADER_KEY]: this.accessToken,
+      'Content-Type': 'application/json',
+    });
+  }
+
+  /**
+   * @param {string} customerId ClientCustomer Id
```
Comment (akartik80) [inline src/client/CartClient.ts:81]:
```suggestion
   * @param {string} customerId Customer ID
```

File: src/client/CollectionClient.ts [hunk 6]
```diff
@@ -16,9 +16,12 @@ export class CollectionClient extends BaseClient {
    * @returns {Promise<CollectionData>}
    */
   async getbyId(id: string): Promise<CollectionData> {
-    const url = `${APIConstants.API_BASE_URL}/collections/${id}`;
+    const params = new Object({
```
Comment (akartik80) [inline src/client/CollectionClient.ts:19]: Please make sure you test all refactoring changes

File: src/client/CollectionClient.ts [hunk 7]
```diff
@@ -34,16 +37,12 @@ export class CollectionClient extends BaseClient {
...
   async getAll(after?: string): Promise<CollectionsData> {
-    const queryParams = new URLSearchParams();
-
-    if (after) {
-      queryParams.append('after', after);
-    }
-
-    const queryParamsString = queryParams.toString() ? `?${queryParams.toString()}` : '';
-    const url = `${APIConstants.API_BASE_URL}/collections${queryParamsString}`;
+    const params = new Object({
```
Comment (akartik80) [inline src/client/CollectionClient.ts:40]:
```suggestion
    const params = {
```

Comment (akartik80) [reply src/client/CollectionClient.ts:40]: Please change everywhere

**Review by akartik80 — CHANGES_REQUESTED** [review]
Take extra caution while writing docs. They are the first impression of who we are

File: src/client/CartClient.ts (hunk 2 above)
Comment (princepatoliya) [reply src/client/CartClient.ts:50]: Okayy

File: src/client/CartClient.ts (hunk 3 above)
Comment (princepatoliya) [reply src/client/CartClient.ts:55]: okay, i'll remmenber that next time

File: src/types/Cart.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
+import { Image } from './Image';
+import { PageInfo } from './PageInfo';
+import { Product, Variant } from './Product';
+
+export enum DiscountType {
+  FIXED_AMOUNT,
+  FIXED_PERCENTAGE,
```
Comment (akartik80) [inline src/types/Cart.ts:7]:
```suggestion
  PERCENTAGE
```

Dhyan se bhai

File: src/types/Cart.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+export type TaxLine = {
+  readonly rate: number;
+  readonly ratePercentage: number;
+  readonly title: string;
+};
+
+export type CartLineItem = {
+  readonly appliedDiscount: Discount;
+  readonly discountedTotal: string;
+  readonly discountedUnitPirce: string;
```
Comment (akartik80) [inline src/types/Cart.ts:32]: typo

File: src/types/Cart.ts [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  readonly id: string;
+  readonly image: Image;
+  readonly name: string;
+  readonly originalTotal: string;
+  readonly originalUnitPrice: string;
+  readonly product: Product;
+  readonly quantity: number;
+  readonly sku: string;
+  readonly taxLines: readonly TaxLine[];
+  readonly texable: boolean;
```
Comment (akartik80) [inline src/types/Cart.ts:42]: Typo. Please do two things:
1. Install any spell checker extension
2. Copy paste properties, always.

Please check all of these. I haven't checked all

**Review by akartik80 — CHANGES_REQUESTED** [review]

**Review by akartik80 — APPROVED** [review]

**Pull request merged**